File: solutions/r/perfect-numbers/2/perfect-numbers.R
sigma <- function(n){
  sum <- 0
  for(k in 1:floor(n/2)){
    if(n%%k==0) sum <- sum + k
  }
  sum
}

number_type <- function(n){
  #stopifnot(n>=1)
  if(n<1) stop("not a natural number")
  if(n==1) return("deficient")
  a <- sigma(n)
  type <- "abundant"
  if (a<n) type <- "deficient"
  if (a==n) type <- "perfect"
  type
}

File: solutions/dart/perfect-numbers/2/lib/perfect_numbers.dart
enum Classification { deficient, abundant, perfect}

class PerfectNumbers {

  int aliquotSum(int n){
    int sum = 1;
    for(var k=2; k< n; k++){
      if(n % k == 0)
        sum += k;
    }
    return sum;
  }
  
  Classification classify(int number){
    if(number<1){
      throw ArgumentError('Not a positive integer');
    } 
    if(number==1)
      return Classification.deficient;
    int aSum = aliquotSum(number);
    if(aSum > number)
      return Classification.abundant;
    else if (aSum == number)
      return Classification.perfect;
    else
      return Classification.deficient;
  }
}
